### DISCUSSION TITLE
Bug in announce example code?

### DISCUSSION THREAD

**Discussion by ThomasFreedman** [markqvist/Reticulum#288] [Help & Questions]
I'm just getting started programming with reticulum, so I may be overlooking something.

I made a few minor mods to the Announce example program (see code below) but it looks like no announcements are being sent. I'm using RNS 0.5.0, Nomadnet 0.3.5 and Sideband 0.5.2. 

I am using a shared instance of rnsd started by a systemd unit, which works fine for both nomadnet & sideband, neither of which see the announcements by this example program. I had the "enable_transport" set False since this is on a laptop, but changing it to True and restarting the rnsd systemd unit didn't resolve the issue. I don't think that shouldn't matter anyway, if all programs are using the same rnsd instance, right?

I launch this program without any command line arguments.

Note my modification to remove the aspect filtering in the example program below. It prints all announcements displayed by nomadnet and sideband. Since this program **does** see the announcements from both of those clients, it's fair to assume the rnsd is setup & working OK. 

```
##########################################################
# This is a slightly modified version of the RNS example #
# that demonstrates setting up announcements & callbacks.#
##########################################################

import argparse
import random
import RNS

APP_NAME = "announce_test"

# We initialize two lists of strings to use as app_data
fruits = ["Peach", "Quince", "Date", "Tangerine", "Pomelo", "Carambola", "Grape"]
noble_gases = ["Helium", "Neon", "Argon", "Krypton", "Xenon", "Radon", "Oganesson"]


# This initialization is executed when the program is started
def program_setup(configpath):
    # We must first initialise Reticulum
    RNS.Reticulum(configpath)

    # Randomly create a new identity for our example
    identity = RNS.Identity()

    # Using the identity we just created, we create two destinations
    # in the "announce_test.announce_sample" application space.
    #
    # Destinations are endpoints in Reticulum, that can be addressed
    # and communicated with. Destinations can also announce their
    # existence, which will let the network know they are reachable
    # and autoomatically create paths to them, from anywhere else
    # in the network.
    destination_1 = RNS.Destination(
        identity,
        RNS.Destination.IN,
        RNS.Destination.SINGLE,
        APP_NAME,
        "announce_sample",
        "fruits"
    )

    destination_2 = RNS.Destination(
        identity,
        RNS.Destination.IN,
        RNS.Destination.SINGLE,
        APP_NAME,
        "announce_sample",
        "noble_gases"
    )

    # We configure the destinations to automatically prove all
    # packets adressed to it. By doing this, RNS will automatically
    # generate a proof for each incoming packet and transmit it
    # back to the sender of that packet. This will let anyone that
    # tries to communicate with the destination know whether their
    # communication was received correctly.
    destination_1.set_proof_strategy(RNS.Destination.PROVE_ALL)
    destination_2.set_proof_strategy(RNS.Destination.PROVE_ALL)

    # We create an announce handler and configure it to only ask for
    # announces from "announce_test.announce_sample.fruits".
    # Try changing the filter and see what happens.
    # REMOVED ASPECT FILTER TO SHOW ALL ANNCOUNCEMENTS ON RNS
    announce_handler = ExampleAnnounceHandler()
#        aspect_filter = APP_NAME + ".announce_sample.fruits"
#    )

    # We register the announce handler with Reticulum
    RNS.Transport.register_announce_handler(announce_handler)

    # Everything's ready!
    # Let's hand over control to the announce loop
    announceLoop(destination_1, destination_2)


def announceLoop(destination_1, destination_2):
    # Let the user know that everything is ready
    print("Announce example running, hit enter to manually send an announce (Ctrl-C to quit)")

    # We enter a loop that runs until the users exits.
    # If the user hits enter, we will announce our server
    # destination on the network, which will let clients
    # know how to create messages directed towards it.
    while True:
        enter = input()
        if enter in ("quit", "exit"): break

        # Randomly select a fruit
        fruit = fruits[random.randint(0, len(fruits) - 1)]

        # Send the announce including the app data
        destination_1.announce(app_data=fruit.encode("utf-8"))
        print(
            "Sent announce from " +
            RNS.prettyhexrep(destination_1.hash) +
            " (" + destination_1.name + ")"
        )

        # Randomly select a noble gas
        noble_gas = noble_gases[random.randint(0, len(noble_gases) - 1)]

        # Send the announce including the app data
        destination_2.announce(app_data=noble_gas.encode("utf-8"))
        print(
            "Sent announce from " +
            RNS.prettyhexrep(destination_2.hash) +
            " (" + destination_2.name + ")"
        )


# We will need to define an announce handler class that
# Reticulum can message when an announce arrives.
class ExampleAnnounceHandler:
    # The initialisation method takes the optional
    # aspect_filter argument. If aspect_filter is set to
    # None, all announces will be passed to the instance.
    # If only some announces are wanted, it can be set to
    # an aspect string.
    def __init__(self, aspect_filter=None):
        self.aspect_filter = aspect_filter

    # This method will be called by Reticulums Transport
    # system when an announce arrives that matches the
    # configured aspect filter. Filters must be specific,
    # and cannot use wildcards.
    @staticmethod
    def received_announce(destination_hash, announced_identity, app_data):
        print(
            "Received an announce from " +
            RNS.prettyhexrep(destination_hash) +
            f" ({announced_identity})"
        )

        if app_data:
            print(
                "The announce contained the following app data: " +
                app_data.decode("utf-8")
            )


##########################################################
#### Program Startup #####################################
##########################################################

# This part of the program gets run at startup,
# and parses input from the user, and then starts
# the desired program mode.
if __name__ == "__main__":
    try:
        parser = argparse.ArgumentParser(
            description="Reticulum example that demonstrates announces and announce handlers"
        )

        parser.add_argument(
            "--config",
            action="store",
            default=None,
            help="path to alternative Reticulum config directory",
            type=str
        )

        args = parser.parse_args()

        if args.config:
            configarg = args.config
        else:
            configarg = None

        program_setup(configarg)

    except KeyboardInterrupt:
        print("")
        exit()

```

**Comment by markqvist** [comment]
I think everything is working as intended. Nomadnet and Sideband will not display announces from the example code, since they are filtering to only receive relevant `lxmf.` destination announces. If you want to see your announces from the example code show up within Sideband or Nomadnet, you will have to create destinations that match what they are filtering for :)

**Comment by ThomasFreedman** [comment]
I didn't know nn or sb was filtering, but it makes sense it is. 

It still doesn't answer why the example code announce handler doesn't see the announces made by the example code. I even removed the aspect filters to see all announces (so I believe). It seems weird you would code an announce handler into the example if it wasn't intended to receive the announces it generates. I only commented out the aspects in the announce handler constructor when I didn't see anything, and when I did I began to see all of the same announcements nn & sb see.

Look at the code. Can you explain why it doesn't see the fruit or noble gas announcements when it does see announcements from nn & sb?  Is there something in the rns config that could explain this?

I created a new file of the announce example from your website, and the only changes made were:
1) added `#!/usr/bin/env python3` to very top of the file 
2) replaced `RNS.log` with `print`

Same results, no announcements show up at all. You say in the description for this example it "builds upon the previous example", but I'm not sure what you mean by that. They appear to be unrelated & independent of each other, aside from the APP_NAME they share.

Also, I just found your PyPi doc about LXMF which says it's built "on top of" reticulum. The line between LXMF & Reticulum is a bit blurry to me. I had to go hunting for LXMF to find what that acronym stood for, as I didn't see it in any of the Reticulum docs I've read. I was actually thinking prior to that discovery LXMF was the low level protocol Reticulum was based on, not the other way around.

**Comment by markqvist** [comment]
> I didn't know nn or sb was filtering, but it makes sense it is.

Yep, in general there is no point in having an application receive announces for anything else than the destinations it cares about. In fact, in most cases you will probably not even need your application as such to receive the announces directly. It will be enough that Reticulum itself handles them. Only if you actually need your application to do something with the event of receiving an announce, or the app data field that it contains, should you need to set up an announce handler.

> Look at the code. Can you explain why it doesn't see the fruit or noble gas announcements when it does see announcements from nn & sb? Is there something in the rns config that could explain this?

Are you sure the code you posted here is the version you are running locally and having trouble with? Caus here it is working completely fine.

Running your code here seems to produce the intended behaviour. When an announce is sent from one instance of the script, the other instance receives it. Works both with both instances running on the same machine (via `rnsd` or not), or if they are on separate systems that are part of the same Reticulum network.

![screen_1](https://user-images.githubusercontent.com/1438012/236313019-766083aa-bf84-4409-9dc0-bf13ebca6bed.png)

So I don't really know what is going wrong, as it is working correctly here.

> Also, I just found your PyPi doc about LXMF which says it's built "on top of" reticulum. The line between LXMF & Reticulum is a bit blurry to me.

You can think of it like this: Reticulum is a networking protocol and stack, that provides the basic networking functionality for other protocols and applications. Like the email protocol and system is build on top of IP+TCP, the LXMF protocol and system is built on top of Reticulum.

Reticulum handles both some very low-level stuff, but also provides much higher-level functionality in one unified stack, with security and better privacy built in. For example Reticulum has the `request()` API built in directly, that lets you do stuff you would normally need a whole host of protocols and pieces of software to carry out (IP+TCP+TLS+HTTP).

**Comment by ThomasFreedman** [comment]
Thanks for your reply Mark. I haven't tried running 2 instances of the program, I'll do that next. I didn't think that was necessary tho, and thought a single instance should be adequate. 

I am thinking that the announce will be the primary mechanism for a publisher, and the announce handler will be used by subscribers. An anounce will be sent for each media file the publisher publishes, with app data being an IPFS hash. That's the basic idea, with reticulum being the ACL comm layer and subscriber qualification and IPFS for the actual media content and associated metadata. 

It's a broadcast model, but the broadcast example isn't suitable since it's unencrypted and only good for locally connected nodes, whereas announcements will traverse throughtout the rns network. If a subscriber pins the hash on their IPFS node, an ack can be sent back to the publisher and the amount of decentralization for that media item tracked. It could also be a gauge of the amount of interest subscribers have for that item. I would like to setup the ack similar to how it works on the messageboard example.

Although rns could be used to transfer the actual media content, I believe IPFS is better suited for large files like videos.   

Right now I'm just trying to get a handle on the basics. Now that I see 2 instances are required of the example, one to send and the other to receive the announcements, I can move on. That just wasn't at all obvious, since both the transmission and handler reside in the same file. One clue tho was the fact the handler is a separate class.

**Comment by ThomasFreedman** [comment]
Well I'll be, 2 instances do work! 

The example would be clearer if the announce generator and the handler were in separate executables, or perhaps you just might add comments to say the handler will not display the announces produced by the same file, or state 2 instances are required. 

I presumed the handler callback was autonomous from the announceLoop and the handler callback was executed by the rns or LXMF layer. So it's still not clear to me why the handler doesn't see the announces sent by the announceLoop of the same file. 

I feel like I need to understand this better. From a practical standpoint it makes sense that the entity making the announcement doesn't need to see it's own announcements, but from a logical standpoint why do they seem to be filtered out?
